arrow-check/src/test/kotlin/arrow/check/gen/GenTest.kt

```diff
@@ -0,0 +1,411 @@
+package arrow.check.gen
+
+import arrow.check.PropertySpec
+import arrow.check.property.PropertyConfig
+import arrow.check.property.PropertyTest
+import arrow.check.property.annotate
+import arrow.check.property.assert
+import arrow.check.property.cover
+import arrow.check.property.coverTable
+import arrow.check.property.property
+import pretty.text
+
+class GenTest : PropertySpec({
+
+    // Functor laws
+
+    // Applicative laws
+
+    // Monad laws
+
+    // Alternative laws (with nullable ?: as orElse)
+
+    // MonadPlus laws
+
+    // MonadReader laws
+
+    // Combinators and builders not covered above
+
+    // Individual generators
+    "Gen.long_ with arbitrary range"(
+        PropertyConfig.default() + PropertyConfig.testLimit(10000)
+    ) {
+        val range = forAll(
```

Comment:
This is what requires the multi-shot trick, correct?

Another alternative being (didn't work out how I hoped 😅 The DSL is powerful).

```
operator fun <A, B> String.invoke(config: PropertyConfig, genA: Gen<A>, genB: Gen<B>, f: (A, B) -> Unit): Unit = TODO()

fun PropertyTest.test(name: String, l: Long, range: Range, size: Long): Unit {
  val (min, max) = range.bounds(currSize)
  assert(l >= min) { "$name: Generated long was smaller than min".text() }
  assert(l < max) { "$name: Generated long was larger than max".text() }
}

val rangeGen = Gen.mapN(
  Gen.long(Long.MIN_VALUE..Long.MAX_VALUE),
  Gen.long(Long.MIN_VALUE..Long.MAX_VALUE)
) { a, b ->
    if (a >= b) b to a
    else a to b
}.map { (min, max) -> Range.constant(min, max) }

val sizeGen = Gen.sized { Gen.just(it) }

"Gen.long_ with arbitrary range"(
  PropertyConfig.default() + PropertyConfig.testLimit(10000),
  sizeGen,
  // :'(
  rangeGen.flatMap { range -> Gen.long_(range).map { Pair(it, range) } }
  Gen.long_(range)
) { currSize, (l, range) ->
  test("long_", l, range, currSize)
}
```

Reply:
See on-going discussion here: https://github.com/arrow-kt/arrow-check/pull/35#discussion_r559571810